pkg/workspace/resource/node.go

```diff
@@ -0,0 +1,297 @@
+func (c *NodeManager) updateWorkspaceStatusIfNeeded(ctx context.Context, wObj *kaitov1beta1.Workspace) error {
+	nodes, err := c.getReadyNodesMatchingLabelSelector(ctx, wObj)
+	if err != nil {
+		klog.ErrorS(err, "failed to get ready nodes for workspace status update", "workspace", klog.KObj(wObj))
+		return fmt.Errorf("failed to get ready nodes for workspace: %w", err)
+	}
+
+	readyNodeNames := make([]string, 0, len(nodes))
+	for _, node := range nodes {
+		readyNodeNames = append(readyNodeNames, node.Name)
+	}
+
+	readyNodeSet := sets.New(readyNodeNames...)
+	currentWorkerNodeSet := sets.New(wObj.Status.WorkerNodes...)
+	needsWorkerNodeUpdate := !readyNodeSet.Equal(currentWorkerNodeSet)
+
+	needsResourceStatusUpdate := true
+	if resourceCondition := meta.FindStatusCondition(wObj.Status.Conditions, string(kaitov1beta1.ConditionTypeResourceStatus)); resourceCondition != nil {
+		if resourceCondition.Status == metav1.ConditionTrue && resourceCondition.Reason == "ResourcesReady" {
+			needsResourceStatusUpdate = false
+		}
+	}
+
+	if !needsWorkerNodeUpdate && !needsResourceStatusUpdate {
+		return nil
+	}
+
+	klog.InfoS("Updating workspace status",
+		"workspace", klog.KObj(wObj),
+		"updateWorkerNodes", needsWorkerNodeUpdate,
+		"updateResourceStatus", needsResourceStatusUpdate,
+		"previousNodes", wObj.Status.WorkerNodes,
+		"newNodes", readyNodeNames)
+
+	return c.updateWorkspaceStatusWithBothUpdates(ctx, &client.ObjectKey{Name: wObj.Name, Namespace: wObj.Namespace},
+		readyNodeNames, needsWorkerNodeUpdate, needsResourceStatusUpdate)
+}
+
+// updateWorkspaceStatusWithBothUpdates updates both WorkerNodes and ResourceStatus condition in a single request
+func (c *NodeManager) updateWorkspaceStatusWithBothUpdates(ctx context.Context, name *client.ObjectKey,
+	workerNodes []string, updateWorkerNodes bool, updateResourceStatus bool) error {
+	return retry.OnError(retry.DefaultRetry,
+		func(err error) bool {
+			return apierrors.IsServiceUnavailable(err) || apierrors.IsServerTimeout(err) || apierrors.IsTooManyRequests(err) || apierrors.IsConflict(err)
+		},
+		func() error {
+			wObj := &kaitov1beta1.Workspace{}
+			if err := c.Client.Get(ctx, *name, wObj); err != nil {
+				if !apierrors.IsNotFound(err) {
+					return err
+				}
+				return nil
+			}
+
+			if updateWorkerNodes {
+				wObj.Status.WorkerNodes = workerNodes
+			}
+
+			if updateResourceStatus {
+				condition := metav1.Condition{
+					Type:               string(kaitov1beta1.ConditionTypeResourceStatus),
+					Status:             metav1.ConditionTrue,
+					Reason:             "ResourcesReady",
+					ObservedGeneration: wObj.GetGeneration(),
+					Message:            "All resources are ready and nodes are available",
+					LastTransitionTime: metav1.Now(),
+				}
+				meta.SetStatusCondition(&wObj.Status.Conditions, condition)
+			}
+
+			return c.Client.Status().Update(ctx, wObj)
```

Comment:
is there possibility that both `updateWorkerNodes` and `updateResourceStatus` could be false? and finally there is still `c.Client.Status().Update(ctx, wObj)` in the end